src/core/utils/chainProviders.ts

```diff
@@ -1,23 +1,104 @@
-import { ethers } from 'ethers'
+import { ethers, Contract } from 'ethers'
 import { Chain } from '../constants/chains'
+import { logger } from './logger'
+import { CustomMulticallJsonRpcProvider } from './CustomMulticallJsonRpcProvider'
 
-const provider = (url: string | undefined, chainId: Chain) =>
-  url ? new ethers.providers.StaticJsonRpcProvider(url, chainId) : undefined
+import { MulticallQueue } from './multicall'
+import multicallAbi from '../../contracts/abis/multicall.json'
+import { Multicall } from '../../contracts'
+import { MULTICALL_ADDRESS } from '../constants/MULTICALL_ADDRESS'
+
+const provider = ({
+  url,
+  chainId,
+  enableMulticall,
+}: {
+  url: string | undefined
+  chainId: Chain
+  enableMulticall: boolean
+}) => {
+  logger.info('Using multicall queue providers')
+  if (!url) return undefined
+
+  if (enableMulticall) {
+    logger.info('Using multicall queue providers')
+
+    const provider = new ethers.providers.StaticJsonRpcProvider(url, chainId)
+
+    // deployed on 100+ chains at address
+    // https://www.multicall3.com/deployments
+    const multicallContract = new Contract(
+      MULTICALL_ADDRESS,
+      multicallAbi,
+      provider,
+    ) as Multicall
+
+    const multicallQueue = new MulticallQueue({
+      flushTimeoutMs: 2,
+      maxBatchSize: 100,
+      multicallContract,
+    })
+
+    return new CustomMulticallJsonRpcProvider({
+      url,
+      network: chainId,
+      multicallQueue,
+    })
+  }
+
+  logger.info('Using standard providers')
+  return new ethers.providers.StaticJsonRpcProvider(url, chainId)
```

Comment:
nit: Since this is the short path, it'd be nicer if the branching was the other way around:
```
if (!enableMulticall) {
  logger.info('Using standard providers')
  return new ethers.providers.StaticJsonRpcProvider(url, chainId)
}

// Longer chunk of code to deal with the multicall approach.
```

Reply:
Yep agreed